drivers/serial/uart_npcx.c

```diff
@@ -0,0 +1,376 @@
+{
+	struct uart_npcx_data *data = DRV_DATA(dev);
+
+	data->user_cb = cb;
+	data->user_data = cb_data;
+}
+
+static void uart_npcx_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct uart_npcx_data *data = DRV_DATA(dev);
+
+	if (data->user_cb) {
+		data->user_cb(data->user_data);
+	}
+}
+
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+/******************************************************************************/
+/* UART driver registration */
+static const struct uart_driver_api uart_npcx_driver_api = {
+	.poll_in = uart_npcx_poll_in,
+	.poll_out = uart_npcx_poll_out,
+	.err_check = uart_npcx_err_check,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = uart_npcx_fifo_fill,
+	.fifo_read = uart_npcx_fifo_read,
+	.irq_tx_enable = uart_npcx_irq_tx_enable,
+	.irq_tx_disable = uart_npcx_irq_tx_disable,
+	.irq_tx_ready = uart_npcx_irq_tx_ready,
+	.irq_tx_complete = uart_npcx_irq_tx_complete,
+	.irq_rx_enable = uart_npcx_irq_rx_enable,
+	.irq_rx_disable = uart_npcx_irq_rx_disable,
+	.irq_rx_ready = uart_npcx_irq_rx_ready,
+	.irq_err_enable = uart_npcx_irq_err_enable,
+	.irq_err_disable = uart_npcx_irq_err_disable,
+	.irq_is_pending = uart_npcx_irq_is_pending,
+	.irq_update = uart_npcx_irq_update,
+	.irq_callback_set = uart_npcx_irq_callback_set,
+#endif	/* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static int uart_npcx_init(struct device *dev)
+{
+	const struct uart_npcx_config *config = DRV_CONFIG(dev);
+	uart_inst_t inst = HAL_INSTANCE(dev);
+	struct device *clk_dev = device_get_binding(NPCX_CLOCK_CONTROL_NAME);
+	uint32_t uart_rate;
+
+	/* Turn on device clock first */
+	if (clock_control_on(clk_dev,
+		(clock_control_subsys_t *) &config->clk_cfg) != 0) {
+		return -EIO;
+	}
+
+	/*
+	 * If apb2's clock is not 15MHz, we need to find the other optimized
+	 * values of UPSR and UBAUD for baud rate 115200.
+	 */
+	if (clock_control_get_rate(clk_dev,
+		(clock_control_subsys_t *) &config->clk_cfg, &uart_rate) < 0) {
+		LOG_ERR("UART clock rate get error.");
+	}
+	__ASSERT(uart_rate == 15000000, "Unsupported apb2 clock for UART!");
+
+	/*
+	 * Fix baud rate to 115200. If this value is modified, please also
+	 * modify the delay in uart_set_pad and uart_reset_default_pad_panic.
+	 */
+	NPCX_UPSR(inst) = 0x38;
+	NPCX_UBAUD(inst) = 0x01;
+
+	/*
+	 * 8-N-1, FIFO enabled.  Must be done after setting
+	 * the divisor for the new divisor to take effect.
+	 */
+	NPCX_UFRS(inst) = 0x00;
+
+	/* Initialize UART FIFO if mode is interrupt driven */
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	/* Enable the UART FIFO mode */
+	SET_BIT(NPCX_UMDSL(inst), NPCX_UMDSL_FIFO_MD);
+
+	/* Disable all all UART tx FIFO interrupts */
+	uart_npcx_dis_all_tx_interrupts(dev);
+
+	/* Clear UART rx FIFO */
+	uart_npcx_clear_rx_fifo(dev);
+
+	/* Configure UART interrupts */
+	config->uconf.irq_config_func(dev);
+#endif
+
+	/* Configure pin-mux for uart device */
+	soc_pinctrl_mux_configure(config->alts_list, config->alts_size, 1);
+
+	return 0;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
```

Comment:
How about having two separate drivers, one for polling and one for interrupt driven, that use a common file for most of their logic? Would that increase code size too much?